### ISSUE TITLE
`jax.ensure_compile_time_eval()` unexpectedly not failed when the computation is dependent on the argument

### ISSUE THREAD

**Issue by ayaka14732** [google/jax#16515]
```python
import jax
jax.config.update('jax_platforms', 'cpu')

import jax.numpy as jnp

@jax.jit
def f(x):
    with jax.ensure_compile_time_eval():
        y = x * 15.  # some random computations
        y = y + 1.
    return y

x = jnp.arange(8).reshape((1, 2, 4))

print(f(x))
# print(f.lower(x).compile().as_text())
```

Expected behavior:

`ConcretizationTypeError` should be raised as the computation cannot be evaluated eagerly (it depends on the input of the function).

Actual behavior:

The program runs normally.

Documentation:

https://jax.readthedocs.io/en/latest/_autosummary/jax.ensure_compile_time_eval.html

**Comment by patrick-kidger** [comment]
I believe this is intended. (Certainly I take advantage of the current behaviour, anyway.) Its seems to be more of a "do evaluation at compile time if possible" flag.
